compiler/rustc_hir_analysis/src/collect.rs

```diff
@@ -444,6 +448,80 @@ impl<'tcx> HirTyLowerer<'tcx> for ItemCtxt<'tcx> {
         self.tcx.at(span).type_param_predicates((self.item_def_id, def_id, assoc_ident))
     }
 
+    #[instrument(level = "debug", skip(self, span), ret)]
+    fn select_inherent_assoc_candidates(
+        &self,
+        span: Span,
+        self_ty: Ty<'tcx>,
+        candidates: &Vec<InherentAssocCandidate>,
+    ) -> (Vec<InherentAssocCandidate>, Vec<FulfillmentError<'tcx>>) {
+        // This is all rather hacky. We attempt to unify impl headers against this self ty
+        // (even though we aren't really in a position to do so) as users of IATs in signatures
+        // are unable to explicitly disambiguate which impl to use other than by specifying the
+        // self type.
+
+        assert!(!self_ty.has_infer());
+        let infcx = self
+            .tcx()
+            .infer_ctxt()
+            .ignoring_regions()
+            // We use the next trait solver as it results in alias equality being deferred
+            // which we can then just ignore. We don't evaluate any goals in this infcx so
+            // the actual solver behaviour doesn't really matter...
+            .with_next_trait_solver(true)
```

Comment:
cc @rust-lang/initiative-trait-system-refactor 

We now use the new solver for equating IAT/IAC self types with the impl self types when *resolving* the IAT. We do this just to make the code a little simpler as we don't have to explicitly fold away unnormalized aliases to inference variables.

This is only done when *resolving* the IAT/IAC, actual wellformedness checks still use the old solver so this is not soundness relevant- no possible IAT resolution could be unsound (only lead to weird errors or weird inference guidance).

This also should only affect code using `feature(inherent_associated_types)` as IATs are unstable, and IACs are only resolved with this logic when they are used in the type system under `feature(min_generic_const_args)`.

Reply:
Actually im dumb we were gonna use a deeprejectctxt for this. shout out to remembering things 2s after posting publicly about them